### PR TITLE
Isolatitude HEALPix padding

### PR DESCRIPTION
**Pull request by yikwill**
# Modulus Pull Request

## Description

This PR adds a new "isolatitude" padding strategy for applying 2D convolutions on the HEALPix grid. The name comes from the fact that this padding strategy will preserve zonally uniform (no east-west variance) signals everywhere except at the poles. The default strategy from Karlbauer et al. (2024) breaks zonal uniformity of incoming signals at the eight intersection points of three HEALPix faces (four such points at 45N and four at 45S) as well as along the seams where north (south) HEALPix faces meet another north (south) HEALPix face. The isolatitude padding scheme is illustrated below for an equatorial and northern hemisphere face.

isolat_padding_E_face
isolat_padding_N_face

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: physicsnemo/models/dlwp_healpix/HEALPixRecUNet.py [hunk 1]
```diff
@@ -68,6 +68,7 @@ def __init__(
         enable_healpixpad: bool = False,
         couplings: list = [],
         residual_prediction: bool = True,
+        hpx_padding_mode: str = 'karlbauer',
```
Comment (akshaysubr) [inline physicsnemo/models/dlwp_healpix/HEALPixRecUNet.py]: Would be good to document this parameter and all possible options in the docstring below since this is the main entrypoint for a user

File: physicsnemo/models/dlwp_healpix_layers/healpix_layers.py [hunk 2]
```diff
@@ -604,7 +880,13 @@ def __init__(self, layer, **kwargs):
                 # or if this needs to be removed
                 layers.append(HEALPixPaddingv2(padding=padding))
```
Comment (akshaysubr) [inline physicsnemo/models/dlwp_healpix_layers/healpix_layers.py]: Shouldn't there be an additional condition in the `if` branch here? Otherwise, if I set `enable_healpixpad=True` and `hpx_padding_mode=isolat`, I'll silently get the `karlbauer` padding right?

File: physicsnemo/models/dlwp_healpix_layers/healpix_layers.py [hunk 3]
```diff
@@ -190,7 +189,12 @@ class HEALPixPadding(th.nn.Module):
     Orientation and arrangement of the HEALPix faces are outlined above.
     """
 
-    def __init__(self, padding: int, enable_nhwc: bool = False):
+    def __init__(
+        self,
+        padding: int,
+        enable_nhwc: bool = False,
+        hpx_padding_mode: str = "karlbauer",
```
Comment (akshaysubr) [inline physicsnemo/models/dlwp_healpix_layers/healpix_layers.py]: Not strictly needed for this PR, but it would be great if this module can be compatible with `torch.compile`. Either in this PR or later, would be good to make sure this compiles without errors and benchmark to see how much of the perf gap to the CUDA version can be made up.

**Review by akshaysubr — CHANGES_REQUESTED** [review]
Looks good to me. Only major comment is about some logic that can silently change the padding method.

File: physicsnemo/models/dlwp_healpix/HEALPixRecUNet.py (hunk 1 above)
Comment (yikwill) [reply physicsnemo/models/dlwp_healpix/HEALPixRecUNet.py]: Good point, added.

File: physicsnemo/models/dlwp_healpix_layers/healpix_layers.py (hunk 2 above)
Comment (yikwill) [reply physicsnemo/models/dlwp_healpix_layers/healpix_layers.py]: Good point. I've set it so that an error is thrown if `enable_healpixpad=True` and `hpx_padding_mode != "karlbauer"` since I think we want it to be clear that e2grid only supports Karlbauer padding for now. In the future when we have an optimized version of isolatitude padding, we can port it to e2grid and remove this check.

File: physicsnemo/models/dlwp_healpix_layers/healpix_layers.py (hunk 3 above)
Comment (yikwill) [reply physicsnemo/models/dlwp_healpix_layers/healpix_layers.py]: I agree that it would be good to optimize this padding strategy. At the moment, I believe it is not compatible with `torch.compile`. I have yet to carefully profile the speed of isolatitude padding vs e2grid and non-e2grid Karlbauer padding. In the near future, I will do this and open a new PR to make this padding `torch.compile` compatible.

**Comment by yikwill** [comment]
Thanks for the review. Addressed the concerns and added the `hpx_padding_mode` option to the non-recurrent unet as well. I will save the `torch.compile` compatibility issue for another PR focused on performance optimization, but I agree that it is a near term priority.

**Review by nathanielcresswellclay — COMMENTED** [review]
Sorry I'm so late to this. At this point we should probably do a rebase (or merge dev) to update this branch and make sure everything is still running smoothly. 

Can you also point me to a config in NV-dlesm so I can run some inference/training trials?

**Comment by yikwill** [comment]
No worries. Just merged dev. To avoid clogging the repo with more configs, here's a small config that you could run to test the padding scheme out.

```
defaults:
  - model: double_conv_256_128_128_couple
  - data: era5_hpx64_30var_coupled_sst_6h_24h_perlmutter
  - trainer: 30var_ws_uncertainty_weighted_adamw
  - _self_
hydra.job.chdir: true # to avoid have logs stepping on one another
experiment_name: isolat_test
output_dir: ./isolat_test
checkpoint_name: last
load_weights_only: false
seed: 0
batch_size: 1
learning_rate: 6e-4
num_workers: 8
port: 29450
master_address: localhost
hydra:
  verbose: false
  run:
    dir: ${output_dir}
model:
  hpx_padding_mode: isolat
  enable_nhwc: true
  enable_healpixpad: false
```

**Comment by yikwill** [comment]
The padding code has undergone a major refactor. Notably,
- The individual padding schemes have been moved to a new file called `healpix_paddings.py`. `healpix_layers.py` now only contains the module which wraps the given layer with hpx padding.
- The `enable_healpixpad` model option is now deprecated in favor of a new option called `hpx_padding_mode`. The available modes are `earth2grid`, `karlbauer`, and `isolatitude`. `hpx_padding_mode=earth2grid` produces the same behavior as `enable_healpixpad=True`. For backwards compatibility, I have left `enable_healpixpad` as an option in the UNets, but added a deprecation warning.
- A new model option `compile_padding` has been added which if true wraps the padding layer with torch compile.

**Comment by akshaysubr** [comment]
@yikwill With this deprecated `enable_healpixpad=True` option, can you also list how an existing training config that uses that option should be changed to still use the fast padding kernels?

**Comment by yikwill** [comment]
@akshaysubr Good idea. I've updated the deprecation warning with instructions on how to set `hpx_padding_mode` to replicate the behavior of the legacy option `enable_healpixpad=True`. I've added code to resolve the the correct `hpx_padding_mode`. If no `hpx_padding_mode` is provided and `enable_healpixpad=True`, the deprecation warning will be thrown and `hpx_padding_mode` will be set to `earth2grid`. If no `hpx_padding_mode` is provided and `enable_healpixpad=False`, the deprecation warning will be thrown and `hpx_padding_mode` will be set to `karlbauer`. Thus, the user will get the same behavior no matter what the value of `enable_healpixpad` is in their config. If `hpx_padding_mode` is explicitly provided, it will be used regardless of the value of `enable_healpixpad`.

Do you think there are any other places/ways to document this deprecation?

File: physicsnemo/metrics/climate/healpix_loss.py [hunk 4]
```diff
@@ -1012,8 +1011,9 @@ def __init__(
         open_dict: dict = {"engine": "zarr"},
         selection_dict: dict = {"channel_c": "land_sea_mask"},
         patch_size: int = 3,
-        use_earth2grid_padding: bool = True,
-        enable_nhwc: bool = True,
+        hpx_padding_mode: str = "earth2grid",
+        enable_nhwc: bool = False,
```
Comment (daviddpruitt) [inline physicsnemo/metrics/climate/healpix_loss.py]: What's the reason for switching this to false? Normally we get some speedup using nhwc.

File: physicsnemo/metrics/climate/healpix_loss.py [hunk 5]
```diff
@@ -1032,10 +1032,12 @@ def __init__(
             dictionary of keyword arguments for xarray.open_dataset. Default is {"channel_c": "land_sea_mask"}.
         patch_size: int
             size of the patch. Default is 3.
-        use_earth2grid_padding: bool
-            whether to use earth2grid hpx padding. Default is False.
+        hpx_padding_mode: str
+            HPX padding scheme to use. Default is "earth2grid".
```
Comment (daviddpruitt) [inline physicsnemo/metrics/climate/healpix_loss.py]: Can you add the list of valid options here or in a config, that way we know what's available and what the options do.

File: physicsnemo/models/dlwp_healpix_layers/healpix_layers.py [hunk 6]
```diff
@@ -15,592 +15,139 @@
...
+                    padding_layer = HEALPixPaddingv2(padding=padding)
+                else:
+                    raise ValueError(
+                        "hpx_padding_mode=earth2grid requires earth2grid import, "
+                        "CUDA, and enable_nhwc=False."
+                    )
+            elif hpx_padding_mode == "karlbauer":
+                padding_layer = HEALPixPadding(padding=padding, enable_nhwc=enable_nhwc)
+            elif hpx_padding_mode == "isolatitude":
+                padding_layer =  HEALPixPaddingIsolatitude(
```
Comment (daviddpruitt) [inline physicsnemo/models/dlwp_healpix_layers/healpix_layers.py]: This should have a check to make sure `nside` is passed in if `isolatitude` padding is selected

File: physicsnemo/metrics/climate/healpix_loss.py [hunk 7]
```diff
@@ -1082,12 +1085,28 @@ def __init__(
         self.diag_mask = th.ones(self.n_members, self.n_members) - th.eye(self.n_members)
         # HEALPix padding module (expects [..., F, H, W])
         if self.patch_radius > 0:
-            if self.use_earth2grid_padding:
+            if self.hpx_padding_mode == "earth2grid":
+                if not have_earth2grid or not th.cuda.is_available() or self.enable_nhwc:
```
Comment (daviddpruitt) [inline physicsnemo/metrics/climate/healpix_loss.py]: Can this check be pushed into `healpix_paddings.py`? It would make more sense there since this is specific to paddings and it would allow other code to run the same check.

File: physicsnemo/metrics/climate/healpix_loss.py (hunk 4 above)
Comment (yikwill) [reply physicsnemo/metrics/climate/healpix_loss.py]: We previously had a check that `enable_nhwc` had to be `False` if `enable_healpixpad=True`. However, there is no reason we couldn't have `earth2grid` padding support channels last. I've added `enable_nhwc` as an option to `HEALPixPaddingv2` and updated the default here to be `True`.

File: physicsnemo/metrics/climate/healpix_loss.py (hunk 5 above)
Comment (yikwill) [reply physicsnemo/metrics/climate/healpix_loss.py]: I've updated the docstring to include the valid options and point the user to `healpix_paddings.py` if they want more info.

File: physicsnemo/models/dlwp_healpix_layers/healpix_layers.py (hunk 6 above)
Comment (yikwill) [reply physicsnemo/models/dlwp_healpix_layers/healpix_layers.py]: Added.

File: physicsnemo/metrics/climate/healpix_loss.py (hunk 7 above)
Comment (yikwill) [reply physicsnemo/metrics/climate/healpix_loss.py]: I added a function `make_hpx_padding_layer` to `healpix_paddings.py` that returns the correct padding layer given `hpx_padding_mode`, `enable_nhwc`, etc. Both `HEALPixLayer` and `PatchedEnergyScoreLoss` now use this helper function to get the right padding layer.

**Review by akshaysubr — APPROVED** [review]
